proxy/client/storage/storage.go

```diff
@@ -0,0 +1,171 @@
+type IStorage interface {
+	GetCurrentPartition() (*Partition, error)
+	RotatePartition() (*Partition, error)
+	InsertLineageEvent(*Partition, string) error
+}
+
+// GetCurrentPartition returns the `Partition` where its `isCurrent` is true if available. Otherwise,
+// sql.ErrNoRows will be returned.
+func (s *Storage) GetCurrentPartition() (*Partition, error) {
+	return getCurrentPartition(s.db)
+}
+
+// RotatePartition rotates the lineage events partitions when the current partition hits the bytes or
+// hours limits. It marks the current partition is done for adding more event records, creates a new
+// lineage events partition table, and creates a new record in partitions table.
+func (s *Storage) RotatePartition() (*Partition, error) {
+	p, err := getCurrentPartition(s.db)
+	if err != nil {
+		return nil, err
+	}
+
+	err = inTransaction(s.db, func(tx *sql.Tx) error {
+		if _, err := tx.Exec(fmt.Sprintf("INSERT INTO %s(event, continue) VALUES('', false)", p.Name)); err != nil {
+			return err
+		}
+
+		if _, err := tx.Exec("UPDATE partitions SET is_current=false WHERE name=?", p.Name); err != nil {
+			return err
+		}
+
+		if err := createPartitionTable(tx); err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return getCurrentPartition(s.db)
+}
+
+func (s *Storage) InsertLineageEvent(p *Partition, lineageEvent string) error {
+	return inTransaction(s.db, func(tx *sql.Tx) error {
+		_, err := tx.Exec(fmt.Sprintf("INSERT INTO %s(event, continue) VALUES(?, true)", p.Name), lineageEvent)
+		if err != nil {
+			return err
+		}
+
+		_, err = tx.Exec("UPDATE partitions SET size=size+? WHERE name=?", len(lineageEvent), p.Name)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// New returns a Storage instance which manages the partitions and checkpoint.
+func New(conf config.Config) (*Storage, error) {
+	db, err := sql.Open("sqlite3", conf.SqliteDatabasePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := initDB(db); err != nil {
+		return nil, err
+	}
+
+	return &Storage{db: db}, nil
+}
+
+func initDB(db *sql.DB) error {
+	_, err := getCurrentPartition(db)
+	if err == sql.ErrNoRows {
+		return inTransaction(db, func(tx *sql.Tx) error {
+			return createPartitionTable(tx)
+		})
+	}
+
+	return err
+}
+
+func getCurrentPartition(db *sql.DB) (*Partition, error) {
+	var (
+		rowID     uint64
+		name      string
+		size      uint64
+		createdAt time.Time
+		isCurrent bool
+	)
+
+	row := db.QueryRow("SELECT rowid, name, size, created_at, is_current FROM partitions WHERE is_current=true")
+	if err := row.Scan(&rowID, &name, &size, &createdAt, &isCurrent); err != nil {
+		return nil, err
+	}
+
+	return &Partition{RowID: rowID, Name: name, Size: size, CreatedAt: createdAt, IsCurrent: isCurrent}, nil
+}
+
+func createPartitionTable(tx *sql.Tx) error {
+	tableName := fmt.Sprintf("lineage_events_partition_%s", generatePartitionSuffix())
```

Comment:
Outside of this method, it's not clear how `lineage_events_partition_*` is used. Can we adds docs to the method?